docs/src/main/mdoc/datatypes/statet.md

```diff
@@ -0,0 +1,244 @@
+---
+layout: docs
+title:  "StateT"
+section: "data"
+source: "core/src/main/scala/cats/data/package.scala"
+scaladoc: "#cats.data.StateT"
+---
+# StateT
+
+`StateT[F[_], S, A]` is a data type used to define state machine
+effectful computations that can be easily composed into a larger one.
+
+The type parameters are:
+- `F[_]` represents the effect in which the computation is performed.
+- `S` represents the underlying state, shared between each step of the
+  state machine.
+- `A` represents the return value.
+
+It can be seen as a way to add the capability to manipulate a shared
+state, to an existing computation in the context of `F`.
+
+This definition could be confusing, but it will become clear after
+learning the `State` monad and by the example below.
+
+## StateT and State Relationship
+
+`StateT` is a monad transformer for
+[State](https://typelevel.org/cats/datatypes/state.html), in
+particular [State](https://typelevel.org/cats/datatypes/state.html) is
+defined as a `StateT` with
+[Eval](https://typelevel.org/cats/datatypes/eval.html) as the effect
+`F`.  Therefore, `StateT` exposes the same methods of
+[State](https://typelevel.org/cats/datatypes/state.html), such as:
+`modify`, `get` and `set`.  Plus additional methods, that handles
+effectful computations, eg: `modifyF`, `setF` and `liftF`.
+
+## Example: Table Reservation System
+
+In this example, we are going to model a *Table Reservation System*. To
+do so, we need to keep track of the tables and the current
+reservations for each of them. The end-user can then try to insert a
+booking for a specific table and time. If such a table is available,
+then the booking is placed and the state is updated, otherwise, an
+error is returned.
+
+To simplify the logic, for each reservation we will just consider a
+single `LocalTime` starting at the beginning of the hour.
+
+Let's start with defining the type synonyms for the state of the
+program and the effect type:
+
+```scala mdoc:silent
+import cats.data.{StateT, NonEmptyList}
+import cats.implicits._
+import java.time.LocalTime
+
+type TRSState = Map[(Int, LocalTime), String]
+type EffectType[A] = Either[Throwable, A]
```

Comment:
I'm not sure about these type aliases. Can we use a `case class` for `TRSState` instead? Maybe call it `Reservations`?

For `EffectType`, I wonder if we use `Either[Throwable, *]` directly in the code. Less sure about this one.

Reply:
I was thinking the same when coding it 😄 
What I can do is to create a `Reservation` case class this way:
```
case class ReservationId(tableNumber: Int, hour: LocalTime)
case class Reservation(id: ReservationId, name: String)
```
and then create a `Reservations` class extending `Map[ReservationId, Reservation]` that can be constructed from `NonEmptyList[Reservation]`. I would like to keep the `Map` capabilities due to the fact that the comparison needs to happen on the key only in order to identify clashes later on, wdyt?

Regarding the `EffectType` to `Either[Throwable, *]`. It can definitely be done, but since it's used multiple times below, I find it more concise to have a type synonym. It's still a matter of taste, if you insist one more time I'll do the change 😛

Reply:
Btw I had another idea for this one, maybe `type ThrowableOr[A] = Either[Throwable, A]`. Actually I wonder why we don't have this alias in cats.

Reply:
> create a `Reservations` class extending `Map[ReservationId, Reservation]`

Hmm, I think this would be rather unusual, like using `Eval` to suspend side-effects. Using a `case class` (and defining extra methods on it) is more idiomatic than extending a `Map` (especially the Scala collections APIs are very complicated, even the collections in cats do not extend them).

I like your case classes 👍

Reply:
Fair, I'll apply such changes ;)